bluesky/utils.py

```diff
@@ -1258,19 +1285,231 @@ def clear(self):
+        :param color: The color of the progress bar as a valid HTML/CSS string
+        :param total: The maximum value of the progress bar, defaults to 1
+        :param value: The current value of the progress bar, optional
+        :return:
+        """
+
+        # Do not draw if this was only created recently, otherwise it
+        # will just flash up and disappear
+        if (time.time() - self.creation_time) < self.delay_draw:
+            return
+
+        container = self.containers[pos]
+        with self.lock:
+            ltext, pbar, rtext = container.children
+            ltext.value = label
+            pbar.layout.display = "flex"
+            rtext.value = meta
+
+            if value is not None:
+                pbar.value = value
+                pbar.max = total or 1
+                pbar.style.bar_color = color
+            else:
+                pbar.layout.display = "none"
+
+    def clear(self) -> None:
+        """
+        Clears all progress bars
+        :return: None
+        """
+
+        with self.lock:
+            for container in self.containers:
+                container.close()
+        hide_leftover_progress_bars()
+
+
+def html_color_from_name(name: str) -> str:
+    """
+    Transforms a string into a unique, bright HTML color
+    (hue varies, saturation and value are fixed).
+    Transformation is pure.
+    :param name: The string to convert
+    :return: A string describing a color, valid in HTML/CSS.
+    """
+
+    hue = int(hashlib.sha1(name.encode("utf-8")).hexdigest(), 16) % 360
+    return f"hsl({hue}, 90%, 55%)"
+
+
+def hide_leftover_progress_bars() -> None:
+    """
+    Sets the style in Jupyterlab to remove leftover whitespace
+    from deleted progress bars. Workaround for:
+    https://github.com/jupyterlab/jupyterlab/issues/7354
+    :return: None
+    """
+
+    display(
+        HTML(
+            """
+        <style>
+            .p-Widget.jp-OutputPrompt.jp-OutputArea-prompt:empty {
+                  padding: 0;
+                  border: 0;
+            }
+        </style>
+    """
+        )
+    )
+
+
+def default_progress_bar(status_objs_or_none) -> ProgressBar:
+    return AsciiProgressBar(status_objs_or_none, delay_draw=0.2)
+
+
+def default_notebook_progress_bar(status_objs_or_none) -> ProgressBar:
+    return NotebookProgressBar(status_objs_or_none, delay_draw=0.2)
+
+
+class ProgressBarManager:
+    pbar_factory: Callable[[Any], ProgressBar]
+    pbar: Optional[ProgressBar]
+
+    def __init__(self,
+                 pbar_factory: Callable[[Any], ProgressBar] = default_progress_bar):
+        """
+        Manages creation and tearing down of progress bars.
+        :param pbar_factory: A function that creates a progress bar given an optional list of status objects
+        :return:
+        """
+
+        self.pbar_factory = pbar_factory
         self.pbar = None
 
+    @classmethod
+    def for_notebook(cls, delay_draw: float = 0.2) -> 'ProgressBarManager':
```

Comment:
Should the `delay_draw` parameter be removed?

Reply:
Hmm, possibly. The purpose of adding a factory function was to make it easier to change in the future. The purpose of this method is to quickly create a suitable manager in your startup script for a notebook.

The constructor currently retains the default behaviour for a similar use case in a terminal. If you call `ProgressBarManager()` the object does exactly what it did before. I could also add a `.for_terminal(delay_draw=0.2)` for consistency.

Aside: Does removing the delay_draw from the constructor also have significant backwards-compatibility implications? If so I could use another ABC with an overridable method instead of passing in a function. I'm not sure I like that though because it is verbose and tightly-coupled.

I welcome your thoughts.

Reply:
Oh, when I last looked at it, the `delay_draw` parameter was accepted but then ignore (not passed through. Now there ia `partial` that passes it through. I am in favor of keeping it here so that it can be used.